apis/v1/httproute_types.go

```diff
@@ -263,6 +263,9 @@ type HTTPRouteRule struct {
 	// invalid, 50 percent of traffic must receive a 500. Implementations may
```

Comment:
@youngnick do you remember if the intent here was "exactly 500" or "5xx"?

Reply:
It was definitely "exactly 500", we logged #1200 to do that.

I've been trying to remember why we moved this to "exactly 500", and I _think_ it was to do with partial validity rules.
There's a bunch of discussion in #1112, and even more in #1211  about it. There's also some discussion on #1511, with @mikemorris' comment https://github.com/kubernetes-sigs/gateway-api/pull/1151#issuecomment-1133127038 being a good summary.

I seem to recall not being confident at the time that we didn't want to overcomplicate the spec. It's already pretty complicated, because we were discussing if "zero endpoints" means "not valid" or not.

Looking back, I _think_ the answer we've landed on is that we treat the references between objects differently to possibly-transient conditions on the proxy  anyway. `ResolvedRefs` is for references.

I don't think we should do this until we've gone back through those discussions and checked that we're not breaking any of the assumptions that we made then - or if we are, then we update other documentation as well to make it clearer.

However, if we can all agree that "zero endpoints" should be considered a _transient_ state that _does not impact the validity of the HTTPRoute_, then returning a 503 _in that case_ is okay.

Like I said, we need to clarify what happens here in the other listed cases for 500 errors.

* What happens when there are multiple BackendRefs and _one_ has no endpoints? As it stands, this update leaves that unclear.
* What happens when _all_ the BackendRefs have no endpoints? (Note that this covers the case where there's only one backend that has no endpoints).

I _think_ the answer should be something like:

- Having no endpoints does _not_ make a HTTPBackendRef invalid in configuration terms
- However, a backend with no endpoints MAY (tbh this might need to be SHOULD or even MUST) be treated as invalid for traffic management purposes and return a 503 error code. This means that, if there are multiple backendRefs:
  - each backendRef _must_ get the correct proportion of traffic, even if that means the proportion of traffic bound for that backendRef all gets a 503. This is to ensure that weighted load balancing failures don't happen silently. (There's a case where you're doing a gradual failover, one of the services gets 503, and you don't notice until you flip the weight to 100 percent on the faulty one that we _have_ to avoid.)
  - if all backendRefs have no endpoints, then all traffic that matches that rule will get a 503.

These are basically the same rules as above for 500s, we're basically making a class of traffic that's "invalid at a traffic level, but not at a config level" by doing this.

Reply:
> These are basically the same rules as above for 500s, we're basically making a class of traffic that's "invalid at a traffic level, but not at a config level" by doing this.

Yeah - that all sounds good

Reply:
> There's a case where you're doing a gradual failover, one of the services gets 503, and you don't notice until you flip the weight to 100 percent on the faulty one that we have to avoid.

Can you elaborate on this a bit more?

Reply:
Agreed with @youngnick's summary - went back to read some of my old comments and this seems to align with my thinking at that time.

Reply:
Do you want me to codify parts of your [comment](https://github.com/kubernetes-sigs/gateway-api/pull/3121#discussion_r1628840064) into the godoc @youngnick ?

Reply:
Yes, I added a suggestion to that effect. Once that's done, this LGTM.